### PR TITLE
feat: add string_literal to module identifier allow list

### PR DESCRIPTION
**Pull request by lucasmarcal-faire**
Add string_literal to module identifier allow list.
This case seems to be pretty rare, and besides the fact that it's not mentioned on ModuleMap specs, [Clang actually accepts that](https://github.com/llvm/llvm-project/blob/dafb90dedcda1ad7b94b0bcdbbe7478f7d0f31f6/clang/lib/Lex/ModuleMap.cpp#L1833).

It's rare, but you can find examples by searching for `module " path:/(^|\/)*\.modulemap$/` on Github

The motivation behind creating this PR is to add support to the swift-llbuild Package build that fails while resolving [this modulemap](https://github.com/swiftlang/swift-llbuild/blob/main/products/libllbuild/include/module.modulemap)

This PR alone doesn't fix swift-llbuild build. Some follow-up PRs will be created to add [SPM include folder handling](https://github.com/swiftlang/swift-package-manager/blob/main/Documentation/Usage.md#creating-c-language-targets)

### CODE REVIEW

**Review by adincebic — APPROVED** [review]

**Comment by cgrindel** [comment]
I converted this to a draft while you are still working on it. When you would like a review, feel free to transition it out of the draft state.

File: swiftpkg/internal/modulemap_parser/collect_module.bzl [hunk 1]
```diff
@@ -91,15 +91,20 @@ def _process_module_tokens(parsed_tokens, prefix_tokens, is_submodule):
...
-    module_id_or_asterisk_token, err = tokens.get(parsed_tokens, 1, count = tlen)
+    module_declaration_identifier_token, err = tokens.get(parsed_tokens, 1, count = tlen)
     if err != None:
         return None, err
     consumed_count += 1
-    if not tokens.is_a(module_id_or_asterisk_token, tts.identifier) and \
-       not tokens.is_a(module_id_or_asterisk_token, tts.operator, operators.asterisk):
+
+    # Check if the module identifier is an identifier, asterisk, or string literal
+    # Examples: module Foo, module *, module "Foo/Bar.h"
```
Comment (cgrindel) [inline swiftpkg/internal/modulemap_parser/collect_module.bzl:100]: If I am reading [the spec](https://clang.llvm.org/docs/Modules.html#id27) properly, a `module-id` must be an `identifier`. I don't see that it allows a literal string. Can you explain why you think string literals should be supported here?

Comment (lucasmarcal-faire) [reply swiftpkg/internal/modulemap_parser/collect_module.bzl]: It turns out that the spec is wrong 🫨 

https://github.com/llvm/llvm-project/blob/dafb90dedcda1ad7b94b0bcdbbe7478f7d0f31f6/clang/lib/Lex/ModuleMap.cpp#L1833

The ModuleMap lexical analysis accepts literal strings

File: swiftpkg/internal/modulemap_parser/collect_module.bzl [hunk 2]
```diff
@@ -91,15 +91,20 @@ def _process_module_tokens(parsed_tokens, prefix_tokens, is_submodule):
...
-    module_id_or_asterisk_token, err = tokens.get(parsed_tokens, 1, count = tlen)
+    module_declaration_identifier_token, err = tokens.get(parsed_tokens, 1, count = tlen)
     if err != None:
         return None, err
     consumed_count += 1
-    if not tokens.is_a(module_id_or_asterisk_token, tts.identifier) and \
-       not tokens.is_a(module_id_or_asterisk_token, tts.operator, operators.asterisk):
+
+    # Check if the module identifier is an identifier, asterisk, or string literal
+    # Examples: module Foo, module *, module "Foo"
```
Comment (cgrindel) [inline swiftpkg/internal/modulemap_parser/collect_module.bzl:100]:
```suggestion
    # Examples: module Foo, module *, module "Foo"
    # The specification does not mention that literal strings are supported. However,
    # canonical implementations of module map parsing do support literal strings.
    # https://github.com/llvm/llvm-project/blob/dafb90dedcda1ad7b94b0bcdbbe7478f7d0f31f6/clang/lib/Lex/ModuleMap.cpp#L1833
```

**Review by cgrindel — APPROVED** [review]
Please incorporate my proposed update to the comment. That should provide enough breadcrumbs to understand why we support string literals.

**Comment by cgrindel** [comment]
@mergifyio queue

**Comment by lucasmarcal-faire** [comment]
I guess I have to fix some tests, let me push a commit with that

**Comment by lucasmarcal-faire** [comment]
@cgrindel Failing tests are now fixed.

**Comment by cgrindel** [comment]
It looks like you need to run `bazel run //:tidy`.

**Comment by lucasmarcal-faire** [comment]
@cgrindel done!

**Comment by cgrindel** [comment]
@mergifyio queue

**Pull request merged**